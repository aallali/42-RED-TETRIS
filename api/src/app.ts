
import * as express from "express";
import * as cors from "cors"
import { Socket } from "socket.io";
import Game from "./Game"

import ROOM from "./Room"
import PLAYER from "./Player"
import { ROOM_MODE } from "./types"


const log = console.log
interface IPlayer {
	id: string;
	nickname: string;
}
interface IRoom {
	id: string;
	name: string;
	players: []
	admin: IPlayer
}
class App {

	public app: express.Application;
	public http: any
	public io: Socket
	private port: number = 4242
	private gameIns: Game
	constructor() {
		this.app = express();
		this.app.use(
			cors({
				origin: "*",
				credentials: true,
			})
		);
		this.http = require("http").Server(this.app);
		this.io = require("socket.io")(this.http, {
			pingInterval: 60000,
			cors: {
				origin: "*",
			},
		});
		this.routes()
		this.initSocketEvents()
		this.gameIns = new Game(this.io)

	}
	/**
	 * 
	 */
	private routes() {
		this.app.get("/", (req: any, res: any) => {
			console.log(Array.from(this.gameIns.players))
			res.send(Array.from(this.gameIns.players))
		});
	}
	private initSocketEvents() {
	}
	public listen() {
		console.clear()
		this.http.listen(this.port, () =>
			console.log(`ðŸš€ Server is running on PORT ${this.port}`)
		);
	}
}

export default new App();
