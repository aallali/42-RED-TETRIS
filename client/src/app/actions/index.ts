import {
	SET_PLAYER, SET_GAME,
	LOGOUT_PLAYER, UPDATE_SCORE,
	SET_PLAYER_ADMIN, ADD_ROW,
	PLAYER_LOST, RESET_STATES,
	BROADCAST_SCORE,
	UNSET_PLAYER_ADMIN
} from "../../reducers/player.reducer";
import { UPDATE_PLAYERS, CLEAR_OPPONENTS } from "../../reducers/opponent.reducer"
import { SET_ERROR, CLEAR_ERROR } from "../../reducers/error.reducer"

import {
	SHIFT_TETRO, MORE_TETROS,
	SET_GAME_MODE, SET_GAME_OVER,
	SET_GAME_STARTED, SET_IN_GAME,
	SET_WINNER, RESET_GAME,
	UPDATE_GAME_MODE,
	LEAVE_GAME,
	SET_GAME_TITLE
} from "../../reducers/game.reducer"

export {
	// Players actions
	SET_PLAYER, SET_GAME,
	LOGOUT_PLAYER,
	UPDATE_SCORE,
	SET_PLAYER_ADMIN,
	ADD_ROW, PLAYER_LOST,
	BROADCAST_SCORE,
	UNSET_PLAYER_ADMIN,
	RESET_STATES,

	// Opponents actions
	UPDATE_PLAYERS,
	CLEAR_OPPONENTS,
	// Error actions
	SET_ERROR, CLEAR_ERROR,

	// game actions
	SHIFT_TETRO, MORE_TETROS,
	SET_GAME_MODE, SET_GAME_OVER,
	SET_GAME_STARTED, SET_IN_GAME,
	SET_WINNER, RESET_GAME,
	UPDATE_GAME_MODE,
	LEAVE_GAME,
	SET_GAME_TITLE
}
