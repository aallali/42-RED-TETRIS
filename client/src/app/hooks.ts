import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import type { RootState, AppDispatch } from './store';
import { useGameStatus } from "./TetrisHooks/useGameStatus"
import { useInterval } from "./TetrisHooks/useInterval"
import { usePlayer } from "./TetrisHooks/usePlayer"
import { useStage } from "./TetrisHooks/useStage"


import io from "socket.io-client";

export const socket = io("http://localhost:4242");
export { useGameStatus, usePlayer, useInterval, useStage }
// Use throughout your app instead of plain `useDispatch` and `useSelector`
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
