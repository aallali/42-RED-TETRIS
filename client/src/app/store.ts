import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import playerReducer from "../reducers/player.reducer";
import stageReducer from "../reducers/stage.reducer"
export const store = configureStore({
  reducer: {
    player: playerReducer,
    stage: stageReducer
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
