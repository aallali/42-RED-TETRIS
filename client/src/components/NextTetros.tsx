import React from "react"
import { useAppSelector } from "../app/hooks";
import Stage from "./TetrisV2/components/Stage/Stage";
import { TETROMINOS } from "../helpers/tetrominos";
import "../pages/playboard.css"
import { getTetros } from "../reducers/game.reducer";


const NextTetros: React.FC = () => {
	const tetros = useAppSelector(getTetros)
	function generateMiniStageFromTetro(t: keyof typeof TETROMINOS) {
		let stage = Array.from(Array(4), () => Array(4).fill([0, 'clear']))
		const nextTetro = TETROMINOS[t].shape
		let pos = { x: 0, y: 0 }
		switch (t) {
			case 'O':
				pos = { x: 1, y: 1 }
				break
			case 'Z':
				pos = { x: 1, y: 1 }
				break
			default:
				break
		}
		nextTetro.forEach((row, y) => {
			row.forEach((value, x) => {
				if (value !== 0) {
					stage[y + pos.y][x + pos.x] = [value, 'merged'];
				}
			});
		});
		return stage
	}
	return (
		<div>
			{/*{tetros.length > 0 ? (<h1 className={"text-xl text-gray-200 font-bold"}>Next Tetros :</h1>): null}*/}
			<div className="inline-blocks  flex   space-x-2">
				{tetros.slice(1, 4).map((l, i) => (<Stage  key={i} stage={generateMiniStageFromTetro(l)} STAGE_HEIGHT={4} STAGE_WIDTH={4} CELL_SIZE={15}/>
				))}
			</div>
		</div>

	)
}
export default NextTetros
