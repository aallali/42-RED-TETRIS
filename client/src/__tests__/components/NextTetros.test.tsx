import renderer from "react-test-renderer";
import NextTetros from "../../components/NextTetros"
import { Provider } from 'react-redux';
import { store } from '../../app/store';

test("App render test ", () => {
	const tree = renderer.create(<Provider store={store}>
		<NextTetros />
	</Provider>).toJSON();
	expect(tree).toMatchSnapshot();
});
