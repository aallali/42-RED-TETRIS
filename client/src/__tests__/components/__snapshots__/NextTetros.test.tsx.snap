// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_ NextTetro Component test Next Tetro snaphot  1`] = `
<div>
  <div
    className="inline-blocks  flex   space-x-2"
  >
    <div
      className="sc-gsDKAQ jyRyaZ"
    >
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL imEiLp"
        color="94, 102, 116"
        type="Z"
      />
      <div
        className="sc-bdvvtL imEiLp"
        color="94, 102, 116"
        type="Z"
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL imEiLp"
        color="94, 102, 116"
        type="Z"
      />
      <div
        className="sc-bdvvtL imEiLp"
        color="94, 102, 116"
        type="Z"
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
    </div>
    <div
      className="sc-gsDKAQ jyRyaZ"
    >
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL imEiLp"
        color="94, 102, 116"
        type="T"
      />
      <div
        className="sc-bdvvtL imEiLp"
        color="94, 102, 116"
        type="T"
      />
      <div
        className="sc-bdvvtL imEiLp"
        color="94, 102, 116"
        type="T"
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL imEiLp"
        color="94, 102, 116"
        type="T"
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
    </div>
    <div
      className="sc-gsDKAQ jyRyaZ"
    >
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL jAbiif"
        color="134, 138, 147"
        type="O"
      />
      <div
        className="sc-bdvvtL jAbiif"
        color="134, 138, 147"
        type="O"
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL jAbiif"
        color="134, 138, 147"
        type="O"
      />
      <div
        className="sc-bdvvtL jAbiif"
        color="134, 138, 147"
        type="O"
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
      <div
        className="sc-bdvvtL hygDQU"
        color="244, 239, 201"
        type={0}
      />
    </div>
  </div>
</div>
`;
